decotools/blob_extraction.py

```diff
@@ -0,0 +1,290 @@
+        M00 = -999
+        subImage = self.get_sub_image(image).transpose()
+        M00 = self.get_raw_moment(subImage, 0, 0)
+        return M00
+
+    def get_covariance(self, image):
+        '''Get the raw moments of the image region inside the bounding box
+           defined by this blob group and calculate the image covariance
+           matrix.'''
+        if self.cov == None:
+            subImage = self.get_sub_image(image).transpose()
+            M00 = self.get_raw_moment(subImage, 0, 0)
+            M10 = self.get_raw_moment(subImage, 1, 0)
+            M01 = self.get_raw_moment(subImage, 0, 1)
+            M11 = self.get_raw_moment(subImage, 1, 1)
+            M20 = self.get_raw_moment(subImage, 2, 0)
+            M02 = self.get_raw_moment(subImage, 0, 2)
+            xbar = M10/M00
+            ybar = M01/M00
+            self.cov = np.vstack([[M20/M00 - xbar*xbar, M11/M00 - xbar*ybar],
+                                  [M11/M00 - xbar*ybar, M02/M00 - ybar*ybar]])
+        return self.cov
+
+    def get_principal_moments(self, image):
+        '''Return the maximum and minimum eigenvalues of the covariance matrix,
+           as well as the angle theta between the maximum eigenvector and the
+           x-axis.'''
+        cov = self.get_covariance(image)
+        u20 = cov[0,0]
+        u11 = cov[0,1]
+        u02 = cov[1,1]
+
+        theta = 0.5 * np.arctan2(2*u11, u20-u02)
+        l1 = 0.5*(u20+u02) + 0.5*np.sqrt(4*u11**2 + (u20-u02)**2)
+        l2 = 0.5*(u20+u02) - 0.5*np.sqrt(4*u11**2 + (u20-u02)**2)
+        return l1, l2, theta
+
+
+def group_blobs(blobs, max_dist):
+    '''Given a list of blobs, group them by distance between the centroids of
+       any two blobs.  If the centroids are more distant than max_dist, create a
+       new blob group.'''
+    n = len(blobs)
+    groups = []
+    if n >= 1:
+        # Single-pass clustering algorithm: make the first blob the nucleus of
+        # a blob group.  Then loop through each blob and add either add it to
+        # this group (depending on the distance measure) or make it the
+        # nucleus of a new blob group
+        bg = BlobGroup()
+        bg.add_blob(blobs[0])
+        groups.append(bg)
+
+        for i in range(1, n):
+            bi = blobs[i]
+            isGrouped = False
+            for group in groups:
+                # Calculate distance measure for a blob and a blob group:
+                # blob just has to be < max_dist from any other blob in the group
+                for bj in group.blobs:
+                    if bi.distance(bj) < max_dist:
+                        group.add_blob(bi)
+                        isGrouped = True
+                        break
+            if not isGrouped:
+                bg = BlobGroup()
+                bg.add_blob(bi)
+                groups.append(bg)
+
+    return groups
+
+
+def extract_blobs(image_file, threshold=20., min_area=5., max_area=1000.,
+                  max_dist=150.):
```

Comment:
It would be useful to have as_grey as a keyword arg for this function so the user can control the conversion